FlinkDeltaSink: Add column stats from the parquet file to delta transaction log

gopik opened the pull request:
Extracts the column level stats from parquet file and add it to the delta log in json format, fixes https://github.com/delta-io/connectors/issues/371

- The implementation is not complete as it may not work with all data types yet. Current implementation supports float/double/long/int/binary. String stats are captured as binary.

- Stats extraction needs to do a file io. Current implementation does this in global committer which is run in single thread. This implementation was done for simplicity, we may need to find an alternative to compute stats in parallel.

gopik commented:
@scottsand-db Anything else I need to do before this can be reviewed?

tdas commented:
@scottsand-db is out on vacation, hence the delay. I will try to take a look until he is back.

gopik commented:
@tdas, thanks for the update.

scottsand-db commented:
Hi @gopik - I'm back from vacation and will take a look this week. Thanks for your patience!

scottsand-db commented:
@kristoffSC - take a look?

scottsand-db commented:
Linking the design doc for future visibility: https://docs.google.com/document/d/1GNujU7XpV2eG7OevFBSZeYKNRM-ZrUx7egb7ti5hAF4/edit

scottsand-db commented on flink/src/main/java/io/delta/flink/sink/internal/writer/stats/DeltaFileStats.java:
```diff
@@ -0,0 +1,246 @@
+    }
+}
+
+class JsonStatInteger extends JsonStat {
```
these couldn't be written using a template? perhaps my java is out of date

gopik replied:
It would be great if we can code this using generics. I tried the following but didn't work:

class JsonStatNumber<T> extends JsonStat {
    Statistics<T> stat
   JsonNode getMin() {
      JsonNode.numberNode(stat.getMin());
  }
}

numberNode impl in JsonNode has overload for each numeric type. But T is not constrained to the types that is allowed by numberNode overloads. There might be some other tricks though to make it compile.

Another approach I'm exploring is, instead of manual serialization to json, if we converted it to first Avro GenericRecord and then serialize it to json. The advantage of GenericRecord is, avro library has bunch of utilities on top of record. For example you get comparable implementation for free which useful for merging the files by primary key order.

scottsand-db commented on flink/src/main/java/io/delta/flink/sink/internal/writer/stats/DeltaFileStats.java:
```diff
@@ -0,0 +1,246 @@
+ * The referenced columns are not delta logical columns but the physical columns within the parquet
+ * file.
+ */
+public class DeltaFileStats {
```
can you include an example of what this looks like? e.g. for a table with schema
```
|-- a: struct
|    |-- b: struct
|    |    |-- c: long
|-- d: int
```

gopik replied:
Added this to the javadoc:

+ * For a table with following schema:
+ *
+ * |-- a: struct
+ * |    |-- b: struct
+ * |    |    |-- c: long
+ * |-- d: int
+ *
+ * parquet stats would be captured as:
+ * recordCount: 1000
+ * a.b.c, LongStatistics{min=10, max=2000, nullCount=0}
+ * d,     IntStatistics{min=5, max=5, nullCount=20}
+ *
+ * This implementation recreates the nested structure required for delta stats which then become:
+ * {
+ *   "numRecords": 1000,
+ *   "minValues": {
+ *     "a": {
+ *       "b": {
+ *         "c": 10
+ *       }
+ *     },
+ *     "d": 5
+ *   },
+ *   "maxValues": {
+ *     "a": {
+ *       "b": {
+ *         "c": 2000
+ *       }
+ *     },
+ *     "d": 5
+ *   },
+ *   "nullCounts": {
+ *     "a": {
+ *       "b": {
+ *         "c": 0
+ *       }
+ *     },
+ *     "d": 20
+ *   }
+ * }

scottsand-db commented on flink/src/main/java/org/apache/flink/streaming/api/functions/sink/filesystem/DeltaPendingFile.java:
```diff
@@ -112,20 +119,31 @@ public LinkedHashMap<String, String> getPartitionSpec() {
         LinkedHashMap<String, String> partitionSpec = this.getPartitionSpec();
+        String stats = null;
         long modificationTime = this.getLastUpdateTime();
         String filePath = PartitionPathUtils.generatePartitionPath(partitionSpec) +
```
isn't `basePath` essentially the same as `PartitionPathUtils.generatePartitionPath(partitionSpec)`, considering that `filePath` = `PartitionPathUtils.generatePartitionPath(partitionSpec)` + `fileName`

gopik replied:
It is, but in the current implementation this is made in one place and passed to all the components using arguments.
Also, it doesn't have to be this if my understanding is correct. It could be any directory prefix if there are no partitions.

scottsand-db replied:
I see. Good point. `basePath` is correct.

scottsand-db commented on flink/src/main/java/org/apache/flink/streaming/api/functions/sink/filesystem/DeltaPendingFile.java:
```diff
@@ -112,20 +119,31 @@ public LinkedHashMap<String, String> getPartitionSpec() {
         String filePath = PartitionPathUtils.generatePartitionPath(partitionSpec) +
             this.getFileName();
+        try {
+            ParquetFileStats parquetStats = ParquetFileStats.readStats(
```
i'm curious what @kristoffSC thinks about perform a read on every write.

kristoffSC replied:
yeah, 
I'm a little bit concern that we will do this from one thread -> Global Committer. 
I'm wondering how painful would be to open 100, 200 etc parquet files and read footer from every one.

Maybe we could have some benchmark that would compare before and after for setup like:
500 or 1k events/s,
10 or 100 writers,
checkpoint every 10 seconds
Test duration 2 minutes.
+ some requirements for columns (number, value range etc).

Also @scottsand-db I'm wondering how Delta gets the statistics? It compute them on a fly or from Parquet file?

gopik replied:
@kristoffSC @scottsand-db I've added details to "Read stats in global committer vs file committer" section in the [doc](https://docs.google.com/document/d/1GNujU7XpV2eG7OevFBSZeYKNRM-ZrUx7egb7ti5hAF4/edit#heading=h.lgucd4d560q9). Please take a look.

We can increase the parallelism by pre computing stats in DeltaFileCommitter so that global committer's toAddFile doesn't wait for a file read.

kristoffSC replied:
I will strongly suggest use `DeltaFileCommitter` over `GlobalCommitter` approach for solution that is based on reading stats back from Parquet file.

gopik replied:
@kristoffSC @scottsand-db I've closed on all the comments except the following:
1. Hide stats impl behind feature flag - Could you point me to an example of any existing flag.
2. Implement stats computation in `DeltaFileCommitter` instead of `GlobalCommitter` - I'm working on this.
3. Benchmarking - do we have an existing benchmarking setup? If so, could you point. me to the docs.

scottsand-db replied:
@kristoffSC can you suggest the best way to add a feature flag for the Sink? It doesn't seem like there's any easy setup for this.

do you suggest we
a) have an `.option` API? on the sink?
b) extract the feature flag from the hadoop conf?

kristoffSC replied:
Hi,
I would recommend to do it via `.option(...)`. And the implementation for `.option(...`) for Delta Sink Builder in my opinion should be implemented in separated branch.

The hadoop conf is needed/used only by delta standalone. There is no interaction between Flink specific classes from Delta connector with hadoop conf and I think we should need to have a good reason to change this.

@gopik
 you can check  `DeltaSourceBuilderBase.java` to see how `.option(...)` was implemented there and do the same thing for `DeltaSinkBuilder`. Heads up, both builder implementations differ in low level details. 

Also please note that .option(...) is overloaded, it accepts boolean, int, long and String as option values. Yours needs to have the same.

As for the feature flag, you need to have a `DeltaSourceOptions.java` like class for Sink and add the flag there as a `DeltaConfigOption` instance. 

A good example to follow would be IGNORE_DELETES option from `DeltaSourceOptions.java`. Having the API for .option(...) you can use the same definition as IGNORE_DELETES has, use same BooleanOptionTypeConverter and it should all be working.

kristoffSC commented:
Hi @gopik 
Just a small question, 
do you have check style configured accordingly to instruction?
https://github.com/delta-io/connectors/tree/master/dev

I think couple things like missing empty lines ect. might be off.

scottsand-db commented on flink/src/main/java/io/delta/flink/sink/internal/writer/stats/ParquetFileStats.java:
```diff
@@ -0,0 +1,77 @@
+    }
+
+    /**
+     * Factory method to read stats given path to parquet file.
```
can you add some more high-level information about how you are reading the stats? full scan? just the footer? etc.

gopik replied:
Added comment.

scottsand-db commented:
@kristoffSC - delta on spark computes the stats while writing out the parquet file.

gopik commented:
> Hi @gopik Just a small question, do you have check style configured accordingly to instruction? https://github.com/delta-io/connectors/tree/master/dev
> 
> I think couple things like missing empty lines ect. might be off.

Yes, I've configured it accordingly. I also ran build/sbt flink/checkstyle and it finished successfully.

kristoffSC commented:
> > Hi @gopik Just a small question, do you have check style configured accordingly to instruction? https://github.com/delta-io/connectors/tree/master/dev
> > I think couple things like missing empty lines ect. might be off.
> 
> Yes, I've configured it accordingly. I also ran build/sbt flink/checkstyle and it finished successfully.

Yeah,
we have noticed that "checkstyle" does not catch all the things and formatter can mess things even more.

The thing I was referring to was missing empty line here:
![image](https://user-images.githubusercontent.com/7932805/183847959-b1f3f845-0f25-4ca0-bf6a-754794dc0214.png)

I have a habit to run code formatter manually before every commit (ctr + alt + L in IntelliJ on windows),
but I had my battles with it, since it can mess the import order sometimes even though I have proper configuration ;)

scottsand-db commented on flink/src/main/java/io/delta/flink/sink/internal/writer/stats/DeltaFileStats.java:
```diff
@@ -0,0 +1,286 @@
+ * a.b.c, LongStatistics{min=10, max=2000, nullCount=0}
+ * d,     IntStatistics{min=5, max=5, nullCount=20}
+ *
+ * This implementation recreates the nested structure required for delta stats which then become:
```
this is great javadoc. but why is it for `JsonStatNoop`? shouldn't this be for `DeltaFileStats`?

gopik replied:
Thanks, it should be for `DeltaFileStats`. moved.

scottsand-db commented on flink/src/main/java/io/delta/flink/sink/internal/writer/stats/DeltaFileStats.java:
```diff
@@ -0,0 +1,286 @@
+/**
+ * Factory class that instantiates appropriate {@link JsonStat} based on the delta {@link DataType}.
+ */
+class JsonStatFactory {
```
can you move all of these `JsonStat...` to `JsonFileStats.java`

gopik replied:
Moved to separate file/class.

scottsand-db commented on flink/src/main/java/io/delta/flink/sink/internal/writer/stats/DeltaFileStats.java:
```diff
@@ -0,0 +1,286 @@
+            throw new RuntimeException(e);
+        }
+    }
+    private JsonNode getStat(Type parquetType, ArrayList<String> path,
```
nit: can you format like:
```java
private JsonNode getStat(
        Type parquetType,
        ArrayList<String> path,
        Function<Statistics<?>,
        JsonNode> toJsonNode) {
```

scottsand-db commented on flink/src/main/java/io/delta/flink/sink/internal/writer/stats/ParquetFileStats.java:
```diff
@@ -0,0 +1,81 @@
+    private Map<ColumnPath, Statistics<?>> load(ParquetMetadata parquetMetadata) {
+        Map<ColumnPath, Statistics<?>> columnStats = new HashMap<>();
+        MessageType parquetSchema = parquetMetadata.getFileMetaData().getSchema();
+        // assert that parquetSchema and deltaSchema match.
```
where is this assert happening?

scottsand-db commented on flink/src/main/java/io/delta/flink/sink/internal/writer/stats/ParquetFileStats.java:
```diff
@@ -0,0 +1,81 @@
+        return rowCount;
+    }
+
+    private Map<ColumnPath, Statistics<?>> load(ParquetMetadata parquetMetadata) {
```
nit: javadoc explaining what you are doing (i.e merging the stats)

scottsand-db commented on flink/src/main/java/io/delta/flink/sink/internal/committer/DeltaGlobalCommitter.java:
```diff
@@ -270,9 +270,11 @@ private void doCommit(OptimisticTransaction transaction,
         for (DeltaCommittable deltaCommittable : committables) {
             DeltaPendingFile deltaPendingFile = deltaCommittable.getDeltaPendingFile();
-            AddFile action = deltaPendingFile.toAddFile();
+            AddFile action = deltaPendingFile.toAddFile(basePath, streamSchema);
```
note that here, `basePath` is `Root path of the DeltaTable`

scottsand-db commented on flink/src/main/java/org/apache/flink/streaming/api/functions/sink/filesystem/DeltaPendingFile.java:
```diff
@@ -112,20 +119,31 @@ public LinkedHashMap<String, String> getPartitionSpec() {
     /**
      * Converts {@link DeltaPendingFile} object to a {@link AddFile} object
      *
+     * @param basePath Base path for the file.
```
this is wrong. this is `Root path of the DeltaTable`

scottsand-db commented on flink/src/main/java/org/apache/flink/streaming/api/functions/sink/filesystem/DeltaPendingFile.java:
```diff
@@ -112,20 +119,31 @@ public LinkedHashMap<String, String> getPartitionSpec() {
             this.getFileName();
+        try {
+            ParquetFileStats parquetStats = ParquetFileStats.readStats(
+                new Path(basePath, filePath).toString());
```
please inline comment that this is the absolute file path.

scottsand-db commented on flink/src/main/java/org/apache/flink/streaming/api/functions/sink/filesystem/DeltaPendingFile.java:
```diff
@@ -112,20 +119,31 @@ public LinkedHashMap<String, String> getPartitionSpec() {
+            stats = deltaStats.toJson();
+        } catch (IOException ioe) {
+            // TODO: If we fail computing stats, we should throw and prevent commit.
+            LOG.error("failed computing stats", ioe);
```
can you wrap this whole stats processing inside of a feature flag? `flink.stats.collect` default true. this value will be provided inside of the `conf: Configuration` when the Flink Sink Builder is constructed.

scottsand-db commented on flink/src/test/java/io/delta/flink/sink/internal/writer/stats/ParquetFileStatsTest.java:
```diff
@@ -0,0 +1,77 @@
+        File resourcesDirectory = new File("src/test/resources");
+        String initialTablePath =
+            resourcesDirectory.getAbsolutePath() + PATH;
+        ParquetFileStats stats = ParquetFileStats.readStats(initialTablePath);
```
this will read all the parquet files, right?

gopik replied:
No, this reads only the file pointed by `PATH` variable. Let me delete other files from the repo.

scottsand-db reviewed (changes requested):
Looking good. Left some comments.

Major requests:
- feature flag
- benchmark this feature to determine the overhead of read-after-write for the stats parsing

SteNicholas commented on flink/src/main/java/io/delta/flink/sink/internal/committer/DeltaGlobalCommitter.java:
```diff
@@ -270,9 +270,11 @@ private void doCommit(OptimisticTransaction transaction,
             " files:" + logFiles
         );
+        StructType streamSchema = SchemaConverter.toDeltaDataType(rowType);
+        LOG.info("Schema: " + streamSchema);
```
Could the schema log move into above `Files to be committed to the Delta table:` log?

SteNicholas commented on flink/src/main/java/io/delta/flink/sink/internal/writer/stats/JsonFileStats.java:
```diff
@@ -0,0 +1,208 @@
+            this.stat = stat;
+        }
+
+        public abstract JsonNode getMin();
```
Does this need to introduce the `getRowCount` interface?

gopik replied:
Did you mean "numRecords"? This is added directly in DeltaFileStats.toJson impl.

SteNicholas replied:
@gopik, right. I mean the `numRecords` for stats. I'm not sure whether it's better to add the `numRecords` in `JsonFileStats`.

gopik replied:
The top level json is assembled in DeltaFileStats.toJson() and numRecords is a top level field (as are minValues, maxValues, nullCounts etc). JsonFileStats is a helper class that maps column types from parquet types to json. numRecords doesn't belong to any column. Could you explain via some code how do add numRecords in JsonFileStats?

SteNicholas commented on flink/src/main/java/org/apache/flink/streaming/api/functions/sink/filesystem/DeltaPendingFile.java:
```diff
@@ -112,20 +119,33 @@ public LinkedHashMap<String, String> getPartitionSpec() {
+            DeltaFileStats deltaStats = new DeltaFileStats(schema, parquetStats);
+            stats = deltaStats.toJson();
+        } catch (IOException ioe) {
+            // TODO: If we fail computing stats, we should throw and prevent commit.
```
Which situation will fail to compute stats?

gopik replied:
This implementation reads stats from the parquet files, so IOException during file reads may prevent stats computation.

SteNicholas commented on flink/src/main/java/org/apache/flink/streaming/api/functions/sink/filesystem/DeltaPendingFile.java:
```diff
@@ -112,20 +119,33 @@ public LinkedHashMap<String, String> getPartitionSpec() {
     /**
      * Converts {@link DeltaPendingFile} object to a {@link AddFile} object
      *
+     * @param basePath Root path of the DeltaTable.
```
```suggestion
     * @param basePath Root path of the delta table.
```

scottsand-db commented:
Hi @gopik - just following up on this. You still need to refactor it so that the parquet footer reading is done inside `DeltaFileCommitter` instead of `DeltaGlobalCommitter`, correct?

gopik commented:
@scottsand-db @kristoffSC I've moved the call to compute stats from Global committer to delta committer. Stats needs 2 inputs, full file path and the schema.
There are 2 approaches -

1. DeltaCommitter has path and schema and passes it to DeltaPendingFile just after commit. DeltaPendingFile.onCommit(filePath, Schema). Would have been simpler, but looks like a lot of coupling between committer and pending file.
2. DeltaPendingFile has filePath and schema, only needs a onCommit invocation to compute stats. Looks cleaner but complex because of additional serialization/deserialization logic.

Please take a look and suggest if this approach seems ok or if you have other ideas.

scottsand-db commented:
Posting for visibility: this PR is currently blocked by https://github.com/delta-io/connectors/pull/425, as we need a way to provide options to the sink API to be able to wrap this feature behind a feature flag.

scottsand-db commented:
Hi @gopik - now that https://github.com/delta-io/connectors/pull/425 has been merged, are you interested in continuing this PR?

gopik commented:
@scottsand-db yes, I'll start working on this from next week.
@kristoffSC : could you take a look at the changes related to serialization due to this change and see if they are ok? Specifically changes in flink/src/main/java/io/delta/flink/sink/internal/committables/DeltaCommittableSerializer.java

scottsand-db commented:
@gopik just following up - any update?

gopik commented:
@scottsand-db I'm waiting for an ack from @kristoffSC on the approach. This change requires changes to serialization for checkpointing.

gopik commented:
@scottsand-db @kristoffSC This change is now ready for review. The implementation is now simplified to avoid any changes to serialization during checkpointing by avoiding any state in DeltaPendingFile. I've also added the config option to compute stats optionally (disabled by default).

vkorukanti commented:
This repo has been deprecated and the code is moved under `connectors` module in https://github.com/delta-io/delta repository. [Here](https://github.com/delta-io/connectors/issues/556#issuecomment-1631207507) are the migration steps to recreate this PR in the new repository location.
